src/evo/specialtx_filter.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2025 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <evo/specialtx_filter.h>
+
+#include <evo/assetlocktx.h>
+#include <evo/providertx.h>
+#include <evo/specialtx.h>
+#include <primitives/transaction.h>
+#include <script/script.h>
+#include <span.h>
+#include <streams.h>
+
+/**
+ * Rationale for Special Transaction Field Extraction:
+ *
+ * This implementation extracts specific fields from Dash special transactions
+ * to maintain parity with the bloom filter implementation (CBloomFilter::CheckSpecialTransactionMatchesAndUpdate).
+ *
+ * The fields extracted are those that SPV clients might need to detect:
+ * - Owner/Voting keys: To track masternode ownership and voting rights
+ * - Payout scripts: To detect payments to specific addresses
+ * - ProTx hashes: To track masternode lifecycle and updates
+ * - Collateral outpoints: To track masternode collateral
+ * - Credit outputs: To track platform-related transactions
+ *
+ * Each transaction type has different fields based on its purpose:
+ * - ProRegTx: All identity and payout fields (initial registration)
+ * - ProUpServTx: ProTx hash and operator payout (service updates)
+ * - ProUpRegTx: ProTx hash, voting key, and payout (ownership updates)
+ * - ProUpRevTx: ProTx hash only (revocation tracking)
+ * - AssetLockTx: Credit output scripts (platform credits)
+ */
+// Helper function to add a script to the filter if it's not empty
+static void AddScriptElement(const CScript& script,
+                            const std::function<void(const std::vector<unsigned char>&)>& addElement)
+{
+    if (!script.empty()) {
+        addElement(std::vector<unsigned char>(script.begin(), script.end()));
+    }
+}
+
+// Helper function to add a hash/key to the filter
+template<typename T>
+static void AddHashElement(const T& hash,
+                          const std::function<void(const std::vector<unsigned char>&)>& addElement)
+{
+    addElement(std::vector<unsigned char>(hash.begin(), hash.end()));
+}
+
+void ExtractSpecialTxFilterElements(const CTransaction& tx,
+                                   const std::function<void(const std::vector<unsigned char>&)>& addElement)
+{
+    if (!tx.HasExtraPayloadField()) {
+        return; // not a special transaction
+    }
+
+    switch(tx.nType) {
+    case TRANSACTION_PROVIDER_REGISTER: {
```

Comment:
could you add a text commentary here for `ExtractSpecialTxFilterElements` and for `CBloomFilter::CheckSpecialTransactionMatchesAndUpdate` saying if you modify one - 2nd should be updated to sync with this one, otherwise it's likely that in future only bloom will be updated?